include/invert_quda.h

```diff
@@ -791,7 +821,7 @@ namespace quda {
 
   class PreconCG : public Solver {
     private:
-      Solver *K;
+      std::unique_ptr<Solver> K;
```

Comment:
Not necessarily a design question we need to solve here, but a `std::unique_ptr` isn't a good solution in all cases---for ex, if we pass in an external preconditioner (such as an MG preconditioner) as opposed to letting the solver create it for us.

See the two different constructors for the `GCR` solver as an example.

Reply:
So we should use `shared_ptr` here maybe? Here in PCG as its current shape `unique_ptr` and `shared_ptr` are effectively the same.

Reply:
I think the issue with a `shared_ptr` is it depends on the `MG` preconditioner (or any external preconditioner) being held in a `shared_ptr` everywhere for reference counting reasons, which isn't the case here (though it is something we should address).

I'm inclined to say we should make it a `shared_ptr` here for forward-thinking reasons, since it's all the same in PCG for now, and we can think about making it more robust for external preconditioners in a later PR. Thoughts?

Reply:
Fair point and I am happy to change to `shared_ptr`.